File: rust/herb-linter/src/rules/herb_disable_comment_missing_rules.rs
use crate::herb_disable::parse_herb_disable_content;

use herb::nodes::ERBContentNode;
use herb::Visitor;

rule_visitor!(MissingRulesVisitor);
define_parser_rule!(
  HerbDisableCommentMissingRulesRule,
  "herb-disable-comment-missing-rules",
  Error,
  MissingRulesVisitor
);

impl Visitor for MissingRulesVisitor {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let opening_tag = match &node.tag_opening {
      Some(token) => &token.value,
      None => return,
    };

    if opening_tag != "<%#" {
      return;
    }

    let content = match &node.content {
      Some(token) => &token.value,
      None => return,
    };

    if parse_herb_disable_content(content).is_some() {
      return;
    }

    let trimmed = content.trim();

    if !trimmed.starts_with("herb:disable") {
      return;
    }

    let after_prefix = trimmed["herb:disable".len()..].trim();

    if !after_prefix.is_empty() {
      return;
    }

    self.add_offense(
      "`herb:disable` comment is missing rule names. Specify `all` or list specific rules to disable.",
      node.location.clone(),
    );
  }
}

File: javascript/packages/linter/src/rules/herb-disable-comment-missing-rules.ts
import { ParserRule } from "../types.js"
import { HerbDisableCommentBaseVisitor } from "./herb-disable-comment-base.js"

import { HERB_DISABLE_DIRECTIVE_KEY } from "../herb-disable-comment-utils.js"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HerbDirectiveNode, ParseResult } from "@herb-tools/core"

class HerbDisableCommentMissingRulesVisitor extends HerbDisableCommentBaseVisitor {
  protected checkHerbDisableComment(node: HerbDirectiveNode, _content: string): void {
    if (node.key?.value !== HERB_DISABLE_DIRECTIVE_KEY) return
    if (node.arguments) return

    this.addOffense(
      `\`herb:disable\` comment is missing rule names. Specify \`all\` or list specific rules to disable.`,
      node.location,
    )
  }
}

export class HerbDisableCommentMissingRulesRule extends ParserRule {
  static ruleName = "herb-disable-comment-missing-rules"
  static introducedIn = this.version("0.8.0")
  static defaultEnabledIn = this.version("0.8.0")

  get parserOptions() {
    return {
      herb_directives: true,
    }
  }

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new HerbDisableCommentMissingRulesVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}
